msgpack.c

```diff
@@ -0,0 +1,112 @@
+/* Pi-hole: A black hole for Internet advertisements
+*  (c) 2017 Pi-hole, LLC (https://pi-hole.net)
+*  Network-wide ad blocking via your own hardware.
+*
+*  FTL Engine
+*  MessagePack serialization
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+
+#include "FTL.h"
+#include "api.h"
+
+void pack_eom(int sock) {
+	// This byte is explicitly never used in the MessagePack spec, so it is perfect to use as an EOM for this API.
+	uint8_t eom = 0xc1;
+	swrite(sock, &eom, sizeof(eom));
+}
+
+void pack_basic(int sock, uint8_t format, void *value, size_t size) {
+	swrite(sock, &format, sizeof(format));
+	swrite(sock, value, size);
+}
+
+uint64_t leToBe64(uint64_t value) {
+	char *ptr = (char *) &value;
+	uint32_t part1, part2;
+
+	// Copy the two halves of the 64 bit input into uint32_t's so we can use htonl
+	memcpy(&part1, ptr, 4);
+	memcpy(&part2, ptr + 4, 4);
+
+	// Flip each half around
+	part1 = htonl(part1);
+	part2 = htonl(part2);
+
+	// Arrange them to form the big-endian version of the original input
+	return (uint64_t) part1 << 32 | part2;
+}
+
+void pack_bool(int sock, bool value) {
+	uint8_t packed = (uint8_t) (value ? 0xc3 : 0xc2);
+	swrite(sock, &packed, sizeof(packed));
+}
+
+void pack_uint8(int sock, uint8_t value) {
+	pack_basic(sock, 0xcc, &value, sizeof(value));
+}
+
+void pack_uint64(int sock, uint64_t value) {
+	uint64_t bigEValue = leToBe64(value);
+	pack_basic(sock, 0xcf, &bigEValue, sizeof(bigEValue));
+}
+
+void pack_int32(int sock, int32_t value) {
+	uint32_t bigEValue = htonl((uint32_t) value);
+	pack_basic(sock, 0xd2, &bigEValue, sizeof(bigEValue));
+}
+
+void pack_int64(int sock, int64_t value) {
+	// Need to use memcpy to do a direct copy without reinterpreting the bytes (making negatives into positives).
+	// It should get optimized away.
+	uint64_t bigEValue;
+	memcpy(&bigEValue, &value, sizeof(bigEValue));
+	bigEValue = leToBe64(bigEValue);
+	pack_basic(sock, 0xd3, &bigEValue, sizeof(bigEValue));
+}
+
+void pack_float(int sock, float value) {
+	// Need to use memcpy to do a direct copy without reinterpreting the bytes. It should get optimized away.
+	uint32_t bigEValue;
+    memcpy(&bigEValue, &value, sizeof(bigEValue));
+    bigEValue = htonl(bigEValue);
+	pack_basic(sock, 0xca, &bigEValue, sizeof(bigEValue));
+}
+
+void pack_fixstr(int sock, char *string) {
+	// Make sure that the length is less than 32
+	size_t length = strlen(string);
+
+	if(length >= 32) {
+		logg("Tried to send a fixstr longer than 31 bytes!");
+		exit(EXIT_FAILURE);
+	}
+
+	uint8_t format = (uint8_t) (0xA0 | length);
+	swrite(sock, &format, sizeof(format));
+	swrite(sock, string, length);
+}
+
+void pack_str32(int sock, char *string) {
+	// Make sure that the length is less than 4294967296
+	size_t length = strlen(string);
+
+	if(length >= 4294967296) {
```

Comment:
```
msgpack.c: In function ‘pack_str32’:
msgpack.c:95:12: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  if(length >= 4294967296) {
```

Reply:
When compiling on a 64 bit system, this warning does not appear. Do you think we should check for this?

Reply:
4294967296 is slightly more than 4 GB, so we really expect strings to become that large?

If you want to keep the check (which is probably, a good idea, test for the length of `size_t`), maybe like this:
```c
if((sizeof(size_t) == 8) && length >= 4294967296)
{ ... }
```

There shouldn't be any performance penalty here, as with any (non-zero) level of optimization the compiled binary won't actually perform the test as the compiler will already eliminate whichever branch is dead at compile time.

Reply:
That code still gives the compiler warning. How about reducing it to:
```c
if(length >= 2147483648) { ... }
```
The compiler does not yell about this, but it is a slightly sloppy way to avoid it.

Reply:
I cannot test this at the moment but there should be a preprocessor constant for the maximum allowed size of `size_t`. Maybe something like
```c
#if __SIZE_MAX__ > 4294967295
if(length >= 4294967296)
    { ... }
#endif
```